PR for comments (#4)
soulshake/demo-cruncher: README.md

# Cruncher

This is a simple demo app showing how to implement queue-based batch processing on Kubernetes, and optionally, cluster autoscaling.

It uses AWS EKS and AWS SQS.

## Architecture overview

For each task to execute, a message is posted in the SQS queue.
A custom controller (`queue-watcher`) watches that queue, and receives
these messages. For each message, `queue-watcher` creates a
Kubernetes Job, then deletes the message from the queue.

Then the Kubernetes Job controller kicks in, and run the requested tasks in Pods.
If enough Pods are created, they will fill up the available cluster capacity,
causing new Pods to be in the "Pending" state.
This will then trigger the cluster autoscaler to add more nodes.

Completed Jobs (both successful and failed) are not deleted.
This makes it possible to detect failed jobs and requeue them
(examples will be provided).

## How to run the demo

### Part 1: the setup

Make sure that you have the required dependencies:

- Terraform
- AWS CLI
- `envsubst` (typically found in package `gettext`)
- `jq`

Set the following environment variables:

- `AWS_ACCESS_KEY_ID`
- `AWS_SECRET_ACCESS_KEY`

> - `AWS_DEFAULT_REGION`

Author: ```suggestion
- `AWS_REGION`
```

- `AWS_ACCOUNT_ID`

If you don't know your account ID, you can see it with `aws sts get-caller-identity`

> (after settings access key and secret access key environment variables).

Author: ```suggestion
(after setting the access key and secret access key environment variables).
```

Create the EKS cluster.

```bash
cd cluster
terraform init
terraform apply -target aws_eks_cluster.current
terraform apply
aws eks update-kubeconfig --name default
cd ..
```

Note: this usually takes at least 10-15 minutes to complete.

Note: the name of the cluster (`default` in the example above) corresponds
to the name of the Terraform workspace. If you want to deploy multiple
clusters, you can change the Terraform workspace, and it should deploy
another set of resources. You will also have to edit the cluster name
in `queue/main.tf`.

Create the SQS queue.

```bash
cd queue
terraform init
terraform apply
cd ..
```

This Terraform configuration creates an SQS queue and an IAM role with
minimally scoped permissions. 

Note: the name of the Terraform workspace will be part of the queue
URL, so if you want to create multiple queues, you can do so by changing
the Terraform workspace. However, if you do that, you will need to adjust
a few other manifests where the that value might be hardcoded.

Set the following environment variable; we will use it in a lot of
other commands below.

```bash

> QUEUE_URL=https://sqs.${AWS_DEFAULT_REGION}.amazonaws.com/${AWS_ACCOUNT_ID}/${TF_WORKSPACE-default}

Author: ```suggestion
QUEUE_URL=https://sqs.${AWS_REGION}.amazonaws.com/${AWS_ACCOUNT_ID}/${TF_WORKSPACE-default}
```
